### PR TITLE
[TACHYON-1419] Check - Check write operations in Master side

### PR DESCRIPTION
**Pull request by dongche**
JIRA: https://tachyon.atlassian.net/browse/TACHYON-1419

Check write operations in `FileSystemMaster`, including create, mkdir, rename, deleteFile.

Generally, when a user, say jack, operates file/dir in a file system, there is a /home/jack/ dir for him, instead of under root ‘/’ directly.

It might be better not to bother Tachyon too much, for avoiding path transformation and admin work. So this patch did this in a simple way:
When a general user jack want to create/rename/delete a file under root, say /fileOfJack, we assume he has write permission under root, but with a limitation that he must be the owner of the file. Then he can operate his own, but cannot for others. This is similar with the sticky bit `t` in file system permission.

@apc999 , @jsimsa , @yupeng9 , @sundapeng

### CODE REVIEW

File: integration-tests/src/test/java/tachyon/client/TachyonFileSystemIntegrationTest.java [hunk 1]
```diff
@@ -94,7 +94,7 @@ public void createFileWithFileAlreadyExistsExceptionTest() throws IOException, T
   @Test
   public void createFileWithInvalidPathExceptionTest() throws IOException, TachyonException {
     mThrown.expect(InvalidPathException.class);
-    mThrown.expectMessage("Path root/testFile1 is invalid.");
+    mThrown.expectMessage("Path root/testFile1 is invalid");
```
Comment (yupeng9) [inline integration-tests/src/test/java/tachyon/client/TachyonFileSystemIntegrationTest.java]: You can use the constant defined in ExceptionMessage

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 2]
```diff
@@ -623,11 +634,19 @@ public int getNumberOfPinnedFiles() {
...
-      throws IOException, FileDoesNotExistException, DirectoryNotEmptyException {
+      throws IOException, FileDoesNotExistException, DirectoryNotEmptyException,
+      AccessControlException {
     MasterContext.getMasterSource().incDeletePathOps(1);
     synchronized (mInodeTree) {
+      TachyonURI path = mInodeTree.getPath(mInodeTree.getInodeById(fileId));
+      try {
+        checkPermission(FileSystemAction.WRITE, path, true);
+      } catch (InvalidPathException e) {
+        // no-op
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: What about printing a log?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -1319,11 +1343,13 @@ public void reportLostFile(long fileId) throws FileDoesNotExistException {
...
    * @throws FileAlreadyCompletedException if the file is already completed
    * @throws IOException if an I/O error occurs
+   * @throws AccessControlException if permission checking fails
    */
   // TODO(jiri): Make it possible to load UFS objects recursively.
   public long loadMetadata(TachyonURI path, boolean recursive)
       throws BlockInfoException, FileAlreadyExistsException, FileDoesNotExistException,
-      InvalidPathException, InvalidFileSizeException, FileAlreadyCompletedException, IOException {
+      InvalidPathException, InvalidFileSizeException, FileAlreadyCompletedException, IOException,
+      AccessControlException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Is the checking not added?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 4]
```diff
@@ -1390,7 +1417,7 @@ private long loadMetadataDirectory(TachyonURI path, boolean recursive)
   }
 
   public boolean mount(TachyonURI tachyonPath, TachyonURI ufsPath)
-      throws FileAlreadyExistsException, InvalidPathException, IOException {
+      throws FileAlreadyExistsException, InvalidPathException, IOException, AccessControlException {
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Is the checking not added? Or `AccessControlException`is not necessary?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 5]
```diff
@@ -1750,6 +1777,63 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
   }
 
   /**
+   * Checks user's permission on a path
+   *
+   * @param action requested {@link FileSystemAction} by user
+   * @param path the path to check permission on
+   * @param checkParent indicates whether to check the path or its parent
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: doc: whether to check its parent?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 6]
```diff
@@ -1750,6 +1777,63 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
...
+   * @throws AccessControlException if permission checking fails
+   * @throws InvalidPathException if the path is invalid
+   */
+  private void checkPermission(FileSystemAction action, TachyonURI path,
+      boolean checkParent) throws AccessControlException, InvalidPathException {
+    // bypasses permission checking if security is not enabled.
+    if (MasterContext.getConf().get(Constants.SECURITY_AUTHENTICATION_TYPE).equals("NOSASL")
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Shall we add a util method checking if security is enabled, similarly to how we check if lineage is enabled?

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 7]
```diff
@@ -1750,6 +1777,63 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
...
+      throw new AccessControlException(
+          ExceptionMessage.PERMISSION_DENIED.getMessage(e.getMessage()));
+    }
+
+    // perform permission check
+    if (checkParent) {
+      if ((fileInfos.size() == 1 && pathComponents.length > 1)
+          || (fileInfos.size() == 2 && pathComponents.length == 2)
+          && action.equals(FileSystemAction.WRITE)) {
+        // When checking Write permission on the parent or ancestor of a path, which is root "/",
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: I don't quite understand this comment. Are you trying handle a special case where the path is root or a level under root? Then you can simply state so.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+
+  @Test
+  public void renameSuccessTest() throws Exception {
+    // rename "/testDir/file" to "/testDir/fileRenamed" for user1
+    verifyRename(TEST_USER_1, TEST_DIR_FILE_URI, "/testDir/fileRenamed");
+  }
+
+  @Test
+  public void renameFailNotByPermissionTest() throws Exception {
+    mThrown.expect(InvalidPathException.class);
```
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Could you also verify the expected message for this test and others in this class?

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+    // rename "/testDir/file" to "/testDir/notExistDir/fileRenamed" for user1
+    // This is permitted by permission checking model.
```
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: not permitted?

**Comment by yupeng9** [comment]
Thanks for adding this checking! I left some comments.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 3 above)
Comment (dongche) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: delegated to `create(..)` in this method body.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 4 above)
Comment (dongche) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: `mount()` invoke `loadMetadataDirectory()`, and then invoke `mkdir()`, where checking happens.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 5 above)
Comment (dongche) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Right, thanks!

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java (hunk 7 above)
Comment (dongche) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Trying to handle a special case when the path is a level under root. I will update the comment for more clarity.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 9 above)
Comment (dongche) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: It is permitted by checking logic, but then failed in rename logic (cannot rename a file to a dst path whose parent does not exist).  
Will add expected exception message and update comment to make it more clear.

**Comment by dongche** [comment]
@yupeng9 thanks for your review!

Patch updates:
* Add checking read operation code into this PR. (https://tachyon.atlassian.net/browse/TACHYON-1418). For read operation, it includes methods like getFileInfo(), getFileInfoList(), getFileBlockInfo(), getFileBlockInfoList(). They take a long value representing file id as input, and this long value is got by getFileId(path). Since the api is changing - make only path is used and file id is abandoned, we will add more permission check in read methods after api is stable. Currently only check getFileId(path) method.

* Address comments, except one that add expected exception message in test. Will address this one in next update.

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java (hunk 8 above)
Comment (dongche) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: Yes. Will add it in all tests in next update. Time is tight today...

File: integration-tests/src/test/java/tachyon/master/JournalIntegrationTest.java [hunk 10]
```diff
@@ -104,7 +105,7 @@ private void deleteFsMasterJournalLogs() throws IOException {
   }
 
   private void addBlockTestUtil(FileInfo fileInfo)
-      throws IOException, InvalidPathException, FileDoesNotExistException {
+      throws IOException, InvalidPathException, FileDoesNotExistException, AccessControlException {
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/master/JournalIntegrationTest.java]: I think that we do not need to enumerate all the exceptions a test can throw ... simple catch Exception.

This applies to all functions in this file.

File: integration-tests/src/test/java/tachyon/master/JournalShutdownIntegrationTest.java [hunk 11]
```diff
@@ -129,7 +130,8 @@ private FileSystemMaster createFsMasterFromJournal() throws IOException {
    * Reproduce the journal and check if the state is correct.
    */
   private void reproduceAndCheckState(int successFiles, int successTables) throws IOException,
-      InvalidPathException, FileDoesNotExistException, TableDoesNotExistException {
+      InvalidPathException, FileDoesNotExistException, TableDoesNotExistException,
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/master/JournalShutdownIntegrationTest.java]: ditto

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 12]
```diff
@@ -501,11 +512,14 @@ private void completeFileFromEntry(CompleteFileEntry entry)
    * @throws InvalidPathException
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: please add missing exception descriptions

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 13]
```diff
@@ -623,11 +637,19 @@ public int getNumberOfPinnedFiles() {
...
-      throws IOException, FileDoesNotExistException, DirectoryNotEmptyException {
+      throws IOException, FileDoesNotExistException, DirectoryNotEmptyException,
+      AccessControlException {
     MasterContext.getMasterSource().incDeletePathOps(1);
     synchronized (mInodeTree) {
+      TachyonURI path = mInodeTree.getPath(mInodeTree.getInodeById(fileId));
+      try {
+        checkPermission(FileSystemAction.WRITE, path, true);
+      } catch (InvalidPathException e) {
+        LOG.warn("Invalid Path {} for checking permission: " + e.getMessage(), path);
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: s/Path/path

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 14]
```diff
@@ -1750,6 +1781,62 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
   }
 
   /**
+   * Checks user's permission on a path
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: end sentence with '.'

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 15]
```diff
@@ -1750,6 +1781,62 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
...
+    }
+
+    // perform permission check
+    if (checkParent) {
+      if ((fileInfos.size() == 1 && pathComponents.length > 1)
+          || (fileInfos.size() == 2 && pathComponents.length == 2)
+          && action.equals(FileSystemAction.WRITE)) {
+        // Handle a special case where the path is a level under root "/" and checking write
+        // permission on it.
+        // we simply assume user has write permission on the root "/",
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: We

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 16]
```diff
@@ -1750,6 +1781,62 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
...
+          ExceptionMessage.PERMISSION_DENIED.getMessage(e.getMessage()));
+    }
+
+    // perform permission check
+    if (checkParent) {
+      if ((fileInfos.size() == 1 && pathComponents.length > 1)
+          || (fileInfos.size() == 2 && pathComponents.length == 2)
+          && action.equals(FileSystemAction.WRITE)) {
+        // Handle a special case where the path is a level under root "/" and checking write
+        // permission on it.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: remove line break

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 17]
```diff
@@ -116,10 +137,12 @@ public static void checkPermission(String user, List<String> groups, FileSystemA
    * @param action requested {@link FileSystemAction} by user
    * @param path the path to check permission on
    * @param fileInfoList file info list of all the inodes retrieved by traversing the path
+   * @param isCheckOwner indicates whether to check the user is the owner of the path
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: s/isCheckOwner/checkOwner

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 18]
```diff
@@ -0,0 +1,443 @@
...
+import tachyon.master.file.options.CreateOptions;
+import tachyon.master.file.options.MkdirOptions;
+import tachyon.master.journal.Journal;
+import tachyon.master.journal.ReadWriteJournal;
+import tachyon.security.authentication.PlainSaslServer;
+import tachyon.security.group.GroupMappingService;
+import tachyon.thrift.FileInfo;
+
+/**
+ * Unit test for {@link tachyon.master.file.FileSystemMaster} when permission check is enabled by
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: import tachyon.master.file.FileSystemMaster and remove the full namespace from the link

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 19]
```diff
@@ -0,0 +1,443 @@
...
+ * configure tachyon.security.authorization.permission.enabled=true.
+ */
+public class FileSystemMasterPermissionCheckTest {
+  private static final String TEST_SUPER_GROUP = "test-supergroup";
+
+  /**
+   * The file structure for testing is:
+   *    /               admin     admin       755
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: use `<code> </code>` or HTML to make sure this shows up properly in javadoc

File: servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java [hunk 20]
```diff
@@ -0,0 +1,443 @@
...
+
+  @Rule
+  public TemporaryFolder mTestFolder = new TemporaryFolder();
+
+  @Rule
+  public ExpectedException mThrown = ExpectedException.none();
+
+  /**
+   * The user and group mappings for testing are:
+   *    admin -> admin
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterPermissionCheckTest.java]: ditto

**Comment by jsimsa** [comment]
@dongche thanks for the PR, left some comments for you

**Comment by dongche** [comment]
@jsimsa thanks for your review. Patch updated to address comments and add exception message in tests

**Comment by dongche** [comment]
@apc999 , @yupeng9 , @jsimsa , patch updated to add https://tachyon.atlassian.net/browse/TACHYON-1422 (Check - Make UI work with permission check). Sorry for submitting several Jiras in one PR, to speed up for catching release cut.

Currently UI contacts with Tachyon cluster without user concept. If enable permission check, one problem is that UI request would get failure because it could not pass permission check in master.
One solution is appending UI request with user, but this is not a effort-less work and need http authentication. This may be another feature.

So a workaround is bypassing permission check if file operation is from UI. Considering UI are just read operation, it would not harm Tachyon at least. This patch bind the file system owner (master login user) to the UI request thread, so as to bypassing permission check. Verified and passed in local cluster test.

**Comment by jsimsa** [comment]
LGTM

**Comment by yupeng9** [comment]
LGTM

**Pull request merged**